src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -325,6 +353,16 @@ List<Binding> getIndirectBindings() {
         List<Binding> getBindings(Class<?> carrier, MemoryLayout layout) {
             TypeClass argumentClass = TypeClass.classifyLayout(layout);
             Binding.Builder bindings = Binding.builder();
+
+            boolean useIntRegsForFloatingPointArgs = useIntRegsForVariadicFloatingPointArgs() && forVariadicFunction && forArguments;
+
+            // Pass HFA arguments using STRUCT_REGISTER when general purpose registers are being
+            // used to pass floating point arguments. If the HFA is too big to pass entirely in
+            // general purpose registers, then pass it as a struct (i.e. as a STRUCT_REFERENCE).
+            if (useIntRegsForFloatingPointArgs && argumentClass == TypeClass.STRUCT_HFA) {
```

Comment:
A way to potentially improve the readability of this section would be to extract the parts that you need for this special `STRUCT_HFA` scenario from the bodies of the case statements for `STRUCT_REGISTER` and `STRUCT_REFERENCE` into separate methods. Then you can move the if statement that is defined here to the body of the `STRUCT_HFA` statement and call those extracted methods as necessary. For example:

```
case STRUCT_HFA: {
    ...
    if (useIntRegsForFloatingPointArgs) {
        if (layout.byteSize() <= maxLayoutSizeOfHFAInGeneralPurposeRegs()) {
            methodExtractedFromStructRegister(); 
       }
       else {
            methodExtractedFromStructReference();
       }
    }
    ...
    break;
}
```

Reply:
I ended up moving this code during a refactoring so this is no longer applicable. See new PR.